pkg/azuredisk/controllerserver.go

```diff
@@ -517,60 +518,152 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
 			return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) can not be negative", start)
 		}
 	}
+	if d.cloud.KubeClient != nil && d.cloud.KubeClient.CoreV1() != nil && d.cloud.KubeClient.CoreV1().PersistentVolumes() != nil {
+		return d.listVolumesWithKubeClient(ctx, start, int(req.MaxEntries))
+	}
+	return d.listVolumesWithoutKubeClient(ctx, start, int(req.MaxEntries))
+}
 
-	disks, derr := d.cloud.DisksClient.ListByResourceGroup(ctx, d.cloud.ResourceGroup)
-	if derr != nil {
-		return nil, status.Error(codes.Internal, fmt.Sprintf("ListVolumes on rg(%s) failed with error: %v", d.cloud.ResourceGroup, derr.Error()))
+// listVolumesWithKubeClient is a helper function for ListVolumes used for when there is an available kubeclient
+func (d *Driver) listVolumesWithKubeClient(ctx context.Context, start, maxEntries int) (*csi.ListVolumesResponse, error) {
+	pvList, err := d.cloud.KubeClient.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "ListVolumes failed while fetching PersistentVolumes List with error: %v", err.Error())
+	}
+
+	// get all resource groups and put them into a sorted slice
+	rgMap := make(map[string]bool)
+	volSet := make(map[string]bool)
+	numVols := 0
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == "disk.csi.azure.com" {
+			diskURI := pv.Spec.CSI.VolumeHandle
+			if err := isValidDiskURI(diskURI); err != nil {
+				return nil, status.Errorf(codes.InvalidArgument, "disk URI(%s) is not valid: %v", diskURI, err)
+			}
+			rg, err := GetResourceGroupFromURI(diskURI)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "could not get resource group from diskURI(%s) with error(%v)", diskURI, err)
+			}
+			volSet[diskURI] = true
+			if _, visited := rgMap[rg]; visited {
+				continue
+			}
+			rgMap[rg] = true
+		}
+		numVols++
 	}
 
-	if start != 0 && start >= len(disks) {
+	if start != 0 && start > numVols {
```

Comment:
Can we do if start!=0 && start>len(pvList.Items) and move this before the above mentioned loop?

Reply:
`pvList.Items` can contain volumes that are not of azure csi driver, so it is inevitable that we loop through each item and confirm that the volume is azure csi disk to get a proper tally of the total volumes `ListVolumes` will return.